include/parallel/parallel_sync.h

```diff
@@ -0,0 +1,341 @@
+ * then act on those answers.
+ *
+ * The \p data map is indexed by processor ids as keys, and for each
+ * processor id in the map there should be a vector of query ids to send.
+ *
+ * Query data which is received from other processors will be operated
+ * on by
+ * gather_data(processor_id_type pid, const std::vector<id> & ids,
+ *             std::vector<datum> & data)
+ *
+ * Answer data which is received from other processors will be operated on by
+ * act_on_data(processor_id_type pid, const std::vector<id> & ids,
+ *             const std::vector<datum> & data);
+ *
+ * No guarantee about operation ordering is made - this function will
+ * attempt to act on data in the order in which it is received.
+ *
+ * All communication and actions are complete when this function
+ * returns.
+ */
+template <typename datum,
+          typename MapToVectors,
+          typename GatherFunctor,
+          typename ActionFunctor>
+void pull_parallel_vector_data(const Communicator & comm,
+                               const MapToVectors & queries,
+                               GatherFunctor & gather_data,
+                               ActionFunctor & act_on_data);
+
+
+//------------------------------------------------------------------------
+// Parallel members
+//
+
+template <typename MapToVectors,
+          typename RequestContainer,
+          typename ActionFunctor>
+void push_parallel_vector_data(const Communicator & comm,
+                               const MapToVectors & data,
+                               RequestContainer & reqs,
+                               ActionFunctor & act_on_data)
+{
+  // This function must be run on all processors at once
+  libmesh_parallel_only(comm);
+
+  processor_id_type num_procs = comm.size();
+
+  // Size of vectors to send to each procesor
+  std::vector<std::size_t> will_send_to(num_procs, 0);
+  processor_id_type num_sends = 0;
+  for (auto & datapair : data)
+    {
+      // Don't try to send anywhere that doesn't exist
+      libmesh_assert_less(datapair.first, num_procs);
+
+      // Don't give us empty vectors to send
+      libmesh_assert_greater(datapair.second.size(), 0);
+
+      will_send_to[datapair.first] = datapair.second.size();
+      num_sends++;
+    }
+
+  // Tell everyone about where everyone will send to
+  comm.alltoall(will_send_to);
+
+  // will_send_to now represents who we'll receive from
+  // give it a good name
+  auto & will_receive_from = will_send_to;
+
+  // Post all of the sends, non-blocking
+  for (auto & datapair : data)
+    {
+      processor_id_type destid = datapair.first;
+      libmesh_assert_less(destid, num_procs);
+      auto & datum = datapair.second;
+      Request sendreq;
+      comm.send(destid, datum, sendreq);
+      reqs.insert(reqs.end(), sendreq);
+    }
+
+  // Post all of the receives, non-blocking
+  std::vector<Request> receive_reqs;
+  std::vector<processor_id_type> receive_procids;
```

Comment:
nit-picky - you could process `will_receive_from` to know how many requests and proc_ids you'll be receiving so you can `reserve()` space for the `push_back()`s